lib/Optimizer/Transforms/DecompositionPatterns.cpp

```diff
@@ -40,6 +41,140 @@ inline Value createDivF(Location loc, Value numerator, double denominator,
   return rewriter.create<arith::DivFOp>(loc, numerator, denominatorValue);
 }
 
+inline bool isAllWires(quake::OperatorInterface op) {
+  for (auto c : op.getControls())
+    if (!c.getType().isa<quake::WireType>())
+      return false;
+  for (auto t : op.getTargets())
+    if (!t.getType().isa<quake::WireType>())
+      return false;
+  return true;
+}
```

Comment:
> Perhaps use `quake::isLinearValueForm` function.

Unless I'm mistaken, I think that will allow `!quake.control` types though, and I don't want to let those types through because they are not fed through to the outputs.

Reply:
Tbh, I'm not familiar with `quake.control`. However, `quake::isLinearValueForm` is what the mapping pass uses to sanity check the ability to map the circuit. Thus, if this function is not a good fit here, then it is likely it shouldn't be a good fit there.

Reply:
> Tbh, I'm not familiar with `quake.control`. However, `quake::isLinearValueForm` is what the mapping pass uses to sanity check the ability to map the circuit. Thus, if this function is not a good fit here, then it is likely it shouldn't be a good fit there.

As a quick test, I tried updating `isAllWires(op)` to simply call `quake::isLinearValueForm(op)`, and that caused the decomposition tests to fail. The reason is because `quake::isLinearValueForm(op)` prohibits both a) an operation's operands from originating from a `quake.unwrap` operation, and b) an operation's results from being sent to a `quake.wrap` operation. That may be a good restriction to enforce during mapping (because that could prevent the mapping pass from comprehending all the operations), but I see no reason to enforce that for this pass. Also, that condition is precisely what happens with the current tests, since the current tests contain `quake.ref` function parameters.

In other words, we are using the `memtoreg` pass to allow the current tests to perform double duty testing (both reference semantics and value semantics), and as an example, that transforms

FROM:
```mlir
func.func @test(%control: !quake.ref, %target: !quake.ref) {
  quake.h [%control] %target : (!quake.ref, !quake.ref) -> ()
  return
}
```
TO:
```mlir
  func.func @test(%arg0: !quake.ref, %arg1: !quake.ref) {
    %0 = quake.unwrap %arg1 : (!quake.ref) -> !quake.wire
    %1 = quake.unwrap %arg0 : (!quake.ref) -> !quake.wire
    %2:2 = quake.h [%1] %0 : (!quake.wire, !quake.wire) -> (!quake.wire, !quake.wire)
    quake.wrap %2#0 to %arg0 : !quake.wire, !quake.ref
    quake.wrap %2#1 to %arg1 : !quake.wire, !quake.ref
    return
  }
```
And `quake::isLinearValueForm(H_Operation)` return false, which is undesirable in my opinion.

If we want to prohibit nearby `quake.unwrap` and `quake.wrap` instructions like the mapping pass does, then I would need to write new tests for all of these decomposition passes, and I (currently) see no need to enforce this tighter restriction on the decomposition pass. What do you think?

Reply:
I don't see how these type changes matter at all with the semantics of the decomposition. They're orthogonal.

What we have is a plurality of ways to express a quantum operation using distinctive types to express different constraints in the IR explicitly and implicitly. The constraints, in fact, do not change. The operation (gate) does not change either.

Decomposition is the replacement of a gate, $a$, with a composition of gates, $a'_0, a'_1, ..., a'_n$. As stated above, the constraints don't change, they are either made explicit or implicit. So the expansion from $a$ to $a'_0, a'_1, ... a'_n$ should simply forward the argument types as presented to $a$ through $a'_0, a'_1, ... a'_n$ to the same end result.

```mlir
   %a:j = quake.a %args ... : (!quake.types ...) -> (!quake.types ...)
```
becomes
```mlir
   %a'_0:j = quake.a'_0 %args ... : (!quake.types ...) -> (!quake.types ...)
   %a'_1:j = quake.a'_1 %args ... : (!quake.types ...) -> (!quake.types ...)
   ...
   %a'_n:j = quake.a'_n %args ... : (!quake.types ...) -> (!quake.types ...)
```

Reply:
This documentation makes it sound like `!quake.control` is never returned as the result of an operation: https://github.com/NVIDIA/cuda-quantum/blob/c4001f66a8f90c0367126b0fdc5df59f24cd8cdf/include/cudaq/Optimizer/Dialect/Quake/QuakeTypes.td#L80.
That seems to conflict with `(!quake.types ...) -> (!quake.types ...)` because the output types do not necessarily have to match the input types, especially when `!quake.control` is used.

Reply:
On the question of "linear types" and the `!quake.control` type, we may be able to fine tune it a bit, but not without really understanding everywhere the test is currently deployed. (In that regard, it's easier to craft a new test for precisely the constraints to be checked than to break other unrelated code in subtle ways.)

`1quake.control` values are SSA values in implementation. However, they are also constrained such that they should only be used within a linear type representation and require "bounding" operations to transform a linear type value to a `control` and convert it back. I believe that was precisely what the test was intended to test for.

Reply:
> Perhaps use `quake::isLinearValueForm` function.

I updated all decompositions to support mixed `quake.wire` / `quake.ref` semantics in 2056fd81c. In that commit, I also updated all decompositions to support mixed `quake.wire` / `quake.ref` / `quake.control` semantics where possible. The function that started this thread (`isAllWires()`) has now been removed, so I'm marking this as resolved. Please feel free to open another thread if we want to continue the discussion on a different topic.